chiselFrontend/src/main/scala/Chisel/Bits.scala

```diff
@@ -84,122 +97,158 @@ sealed abstract class Bits(dirArg: Direction, width: Width, override val litArg:
     * myBits(1,0) => 0b01  // extracts the two least significant bits
     * }}}
     */
-  final def apply(x: Int, y: Int): UInt = {
+  final def apply(x: Int, y: Int): UInt = macro SourceInfoTransform.xyArg
+
+  final def do_apply(x: Int, y: Int)(implicit sourceInfo: SourceInfo): UInt = {
     if (x < y || y < 0) {
       Builder.error(s"Invalid bit range ($x,$y)")
     }
     val w = x - y + 1
     if (isLit()) {
       UInt((litValue >> y) & ((BigInt(1) << w) - 1), w)
     } else {
-      pushOp(DefPrim(UInt(width = w), BitsExtractOp, this.ref, ILit(x), ILit(y)))
+      pushOp(DefPrim(sourceInfo, UInt(width = w), BitsExtractOp, this.ref, ILit(x), ILit(y)))
     }
   }
 
   // REVIEW TODO: again, is this necessary? Or just have this and use implicits?
-  final def apply(x: BigInt, y: BigInt): UInt = apply(x.toInt, y.toInt)
-
-  private[Chisel] def unop[T <: Data](dest: T, op: PrimOp): T =
-    pushOp(DefPrim(dest, op, this.ref))
-  private[Chisel] def binop[T <: Data](dest: T, op: PrimOp, other: BigInt): T =
-    pushOp(DefPrim(dest, op, this.ref, ILit(other)))
-  private[Chisel] def binop[T <: Data](dest: T, op: PrimOp, other: Bits): T =
-    pushOp(DefPrim(dest, op, this.ref, other.ref))
-  private[Chisel] def compop(op: PrimOp, other: Bits): Bool =
-    pushOp(DefPrim(Bool(), op, this.ref, other.ref))
-  private[Chisel] def redop(op: PrimOp): Bool =
-    pushOp(DefPrim(Bool(), op, this.ref))
+  final def apply(x: BigInt, y: BigInt): UInt = macro SourceInfoTransform.xyArg
+
+  final def do_apply(x: BigInt, y: BigInt)(implicit sourceInfo: SourceInfo): UInt =
+    apply(x.toInt, y.toInt)
+
+  private[Chisel] def unop[T <: Data](sourceInfo: SourceInfo, dest: T, op: PrimOp): T =
+    pushOp(DefPrim(sourceInfo, dest, op, this.ref))
+  private[Chisel] def binop[T <: Data](sourceInfo: SourceInfo, dest: T, op: PrimOp, other: BigInt): T =
+    pushOp(DefPrim(sourceInfo, dest, op, this.ref, ILit(other)))
+  private[Chisel] def binop[T <: Data](sourceInfo: SourceInfo, dest: T, op: PrimOp, other: Bits): T =
+    pushOp(DefPrim(sourceInfo, dest, op, this.ref, other.ref))
+
+  private[Chisel] def compop(sourceInfo: SourceInfo, op: PrimOp, other: Bits): Bool =
+    pushOp(DefPrim(sourceInfo, Bool(), op, this.ref, other.ref))
+  private[Chisel] def redop(sourceInfo: SourceInfo, op: PrimOp): Bool =
+    pushOp(DefPrim(sourceInfo, Bool(), op, this.ref))
 
   /** Returns this wire zero padded up to the specified width.
     *
     * @note for SInts only, this does sign extension
     */
-  def pad (other: Int): this.type = binop(cloneTypeWidth(this.width max Width(other)), PadOp, other)
+  def pad(that: Int): this.type = macro SourceInfoTransform.thatArg
```

Comment:
Attempted to add a `type Self <: (this class)` member to every class from `Data`, and ended up with all kinds of errors where the typechecker couldn't infer that `(something).Self` was the same type, so `.asInstanceOf[...]` needs to be spammed everywhere. I think given the amount of stuff that would need to be updated (and whether the `.asInstanceOf[...]` spam is a good tradeoff), this really belongs in another issue / pull request. You should file a bug.

This generally stems from chained calls, like
```scala
def flip(): Self = cloneType.overrideDirection(_.flip, !_)
```
which gives the error
```
[error]  found   : Data.this.Self#Self
[error]  required: Data.this.Self
```